### PR TITLE
Add tag(s) DNS lookup time out.

### LINKED ISSUE #24: tag on failure
after retrying N times, we should tag the event like we do for grok, time, and other filters

### LINKED ISSUE #47: FR: Option to disable logging DNS lookup failures
Hi,

I noticed in my Logstash (v6.4.3) logs many many instances of WARN level messages detailing all the times an IP address could not be reversed to a hostname in my pipeline that leverages logstash-filter-dns reverse lookup. Certainly this information is good to know and good default behavior but in my scenario it's fine if an IP address has no reverse name (PTR record) and I'd like the option to disable logging these reverse lookup failures.

Perhaps adding a Boolean setting called something like 'log_failures' that defaults to 'true' is the way to go?

Or if this is already controllable via Logstash's Log4J properties file, then please update logstash-filter-dns documentation (https://www.elastic.co/guide/en/logstash/current/plugins-filters-dns.html) to explain what to modify in the L4J file to accomplish this without disabling other WARN level messages.

Thanks,
Patrick

### PR DESCRIPTION
**Pull request by mashhurs**
### Description
We had feature requests (#24, #47) to support the tag on DNS look up timeouts and it seems we had a contribution ([PR](https://github.com/logstash-plugins/logstash-filter-dns/pull/53)) but author didn't follow up to end.

This change covers followings:
- adds a tag on timeout, defaults to `_dnstimeout`
- adds `warn_on_timeout` config to tune the log level when dns time out occurs

This PR intends to cover remain works of [original PR](https://github.com/logstash-plugins/logstash-filter-dns/pull/53).

### Closes
Closes #24
Closes #47

### CODE REVIEW

File: docs/index.asciidoc [hunk 1]
```diff
@@ -173,6 +174,13 @@ Reverse resolve one or more fields.
 
 `resolv` calls will be wrapped in a timeout instance
 
+[id="plugins-{type}s-{plugin}-tags_on_timeout"]
+===== `tags_on_timeout`
```
Comment (yaauie) [inline docs/index.asciidoc:178]: prior art in other plugins is `tag_on_timeout` (see: Grok Filter)

File: lib/logstash/filters/dns.rb [hunk 2]
```diff
@@ -86,6 +86,9 @@ class LogStash::Filters::DNS < LogStash::Filters::Base
   # Use custom hosts file(s). For example: `["/var/db/my_custom_hosts"]`
   config :hostsfile, :validate => :array
 
+  # Tag(s) to apply if a DNS lookup times out. Defaults to `["_dnstimeout"]`.
+  config :tags_on_timeout, :validate => :array, :default => ["_dnstimeout"]
```
Comment (yaauie) [inline lib/logstash/filters/dns.rb:90]: We get a little more safety by using ` :validate => :string, :list => true`, which allows either a single string or an array of strings, since each entry will be validated as a string. After validation, regardless of input, we end up with an array of strings.

```suggestion
  config :tag_on_timeout, :validate => :string, :list => true, :default => ["_dnstimeout"]
```

Comment (mashhurs) [reply lib/logstash/filters/dns.rb]: Thanks @yaauie ,
TIL.

File: docs/index.asciidoc (hunk 1 above)
Comment (mashhurs) [reply docs/index.asciidoc]: I have changed back to `tag_on_timeout` aligning with historical content.

File: docs/index.asciidoc [hunk 3]
```diff
@@ -59,6 +59,7 @@ This plugin supports the following configuration options plus the <<plugins-{typ
 | <<plugins-{type}s-{plugin}-resolve>> |<<array,array>>|No
 | <<plugins-{type}s-{plugin}-reverse>> |<<array,array>>|No
 | <<plugins-{type}s-{plugin}-timeout>> |<<number,number>>|No
+| <<plugins-{type}s-{plugin}-tag_on_timeout>> |<<array,array>>|No
```
Comment (jsvd) [inline docs/index.asciidoc:62]: just keeping the list sorted:

```suggestion
| <<plugins-{type}s-{plugin}-tag_on_timeout>> |<<array,array>>|No
| <<plugins-{type}s-{plugin}-timeout>> |<<number,number>>|No
```

File: lib/logstash/filters/dns.rb [hunk 4]
```diff
@@ -219,8 +222,9 @@ def resolve(event)
         end
       rescue Resolv::ResolvTimeout
         @failed_cache[raw] = true if @failed_cache
-        @logger.warn("DNS: timeout on resolving the hostname.",
+        @logger.info("DNS: timeout on resolving the hostname.",
```
Comment (yaauie) [inline lib/logstash/filters/dns.rb:225]: While we have a _request_ to change the logging level, it seems to me that this may break in-the-wild configs that _rely_ on the current logging level. Can/should we make this behaviour configurable?

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,6 @@
+## 3.2.0
+  - Support adding tag(s) on DNS lookup times out, defaults to `["_dnstimeout"]` [#67](https://github.com/logstash-plugins/logstash-filter-dns/pull/67)
```
Comment (yaauie) [inline CHANGELOG.md:2]: change to logging level not referenced in the changelog

File: lib/logstash/filters/dns.rb (hunk 4 above)
Comment (yaauie) [reply lib/logstash/filters/dns.rb:225]: additionally this change is not covered in the specs.

File: CHANGELOG.md (hunk 5 above)
Comment (mashhurs) [reply CHANGELOG.md:2]: Thanks, updated.

File: lib/logstash/filters/dns.rb (hunk 4 above)
Comment (mashhurs) [reply lib/logstash/filters/dns.rb]: Added `warn_on_timeout` flag to provide flexibility. Also covered the unit test cases.

File: lib/logstash/filters/dns.rb [hunk 6]
```diff
@@ -305,8 +317,13 @@ def reverse(event)
...
       rescue Resolv::ResolvTimeout
         @failed_cache[raw] = true if @failed_cache
-        @logger.warn("DNS: timeout on resolving address.",
-                      :field => field, :value => raw)
+        log_message = "DNS: timeout on resolving address."
+        if @warn_on_timeout
+          @logger.warn(log_message, :field => field, :value => raw)
+        else
+          @logger.info(log_message, :field => field, :value => raw)
+        end
```
Comment (roaksoax) [inline lib/logstash/filters/dns.rb:325]: Same as the above.

File: lib/logstash/filters/dns.rb [hunk 7]
```diff
@@ -219,8 +226,13 @@ def resolve(event)
...
       rescue Resolv::ResolvTimeout
         @failed_cache[raw] = true if @failed_cache
-        @logger.warn("DNS: timeout on resolving the hostname.",
-                      :field => field, :value => raw)
+        log_message = "DNS: timeout on resolving the hostname."
+        if @warn_on_timeout
+          @logger.warn(log_message, :field => field, :value => raw)
+        else
+          @logger.info(log_message, :field => field, :value => raw)
+        end
```
Comment (roaksoax) [inline lib/logstash/filters/dns.rb:234]: it seems this change tries to address t[his issue](https://github.com/logstash-plugins/logstash-filter-dns/issues/47).

The requested feature is asking for a wait to _**disable**_ logging. It is NOT asking for a way to change this logging to from WARN to INFO. Can you please clarify why is this change relevant to the attached issue?

**Review by roaksoax — CHANGES_REQUESTED** [review]

File: lib/logstash/filters/dns.rb (hunk 7 above)
Comment (roaksoax) [reply lib/logstash/filters/dns.rb:234]: another option is to make this DEBUG as the original PR.

Comment (mashhurs) [reply lib/logstash/filters/dns.rb]: Long history and it is true that now the logic is adding _extra noise_ :( and doesn't solve the user need.
Stepping back and analyzing, I agree with the _debug_ approach as we added the tag. Users will have two granular options: receiving with tag and debug logs.
Please take a look at the final approach and provide your comments.
Thanks so much for catching this up!

File: lib/logstash/filters/dns.rb [hunk 8]
```diff
@@ -219,8 +222,9 @@ def resolve(event)
         end
       rescue Resolv::ResolvTimeout
         @failed_cache[raw] = true if @failed_cache
-        @logger.warn("DNS: timeout on resolving the hostname.",
+        @logger.debug("DNS: timeout on resolving the hostname.",
```
Comment (roaksoax) [inline lib/logstash/filters/dns.rb:225]: Ok I think it is fair to say that changing the log level is not a breaking change, and we need to re-evalute in the future how we handle logging for event processing, so I'm fine leaving this as debug.

**Review by roaksoax — APPROVED** [review]
